File: Day-02/Python/main.py
import sys

s, t = "", ""
d = {
    (0, 2): 1,
    (1, 1): 2,
    (1, 2): 3,
    (1, 3): 4,
    (2, 0): 5,
    (2, 1): 6,
    (2, 2): 7,
    (2, 3): 8,
    (2, 4): 9,
    (3, 1): 'A',
    (3, 2): 'B',
    (3, 3): 'C',
    (4, 2): 'D'
}
for line in sys.stdin:
    r1, c1, r2, c2 = 1, 1, 2, 0
    for m in line.strip():
        if m == 'U':
            r1 = max(0, r1 - 1)
            if r2 != abs(c2 - 2):
                r2 -= 1
        elif m == 'D':
            r1 = min(2, r1 + 1)
            if r2 != 4 - abs(c2 - 2):
                r2 += 1
        elif m == 'L':
            c1 = max(0, c1 - 1)
            if c2 != abs(r2 - 2):
                c2 -= 1
        else:
            c1 = min(2, c1 + 1)
            if c2 != 4 - abs(r2 - 2):
                c2 += 1
    s += str(3 * r1 + c1 + 1)
    t += str(d[(r2, c2)])
print("Part 1:", s)
print("Part 2:", t)

File: Day-02/C++/main.cpp
#include <bits/stdc++.h>
#include <string>
#include <map>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false); 
    cin.tie(NULL);

    map<int, string> d;
    d.insert(pair<int, string>(2, "1"));
    d.insert(pair<int, string>(11, "2"));
    d.insert(pair<int, string>(12, "3"));
    d.insert(pair<int, string>(13, "4"));
    d.insert(pair<int, string>(20, "5"));
    d.insert(pair<int, string>(21, "6"));
    d.insert(pair<int, string>(22, "7"));
    d.insert(pair<int, string>(23, "8"));
    d.insert(pair<int, string>(24, "9"));
    d.insert(pair<int, string>(31, "A"));
    d.insert(pair<int, string>(32, "B"));
    d.insert(pair<int, string>(33, "C"));
    d.insert(pair<int, string>(42, "D"));

    string line, s = "", t = "";
    while (cin >> line) {
        int r1 = 1, c1 = 1, r2 = 2, c2 = 0;
        for (int i = 0; i < line.length(); i++) {
            if (line[i] == 'U') {
                r1 = max(0, r1 - 1);
                if (r2 != abs(c2 - 2)) {
                    r2 -= 1;
                }
            } else if (line[i] == 'D') {
                r1 = min(2, r1 + 1);
                if (r2 != 4 - abs(c2 - 2)) {
                    r2 += 1;
                }
            } else if (line[i] == 'L') {
                c1 = max(0, c1 - 1);
                if (c2 != abs(r2 - 2)) {
                    c2 -= 1;
                }
            } else {
                c1 = min(2, c1 + 1);
                if (c2 != 4 - abs(r2 - 2)) {
                    c2 += 1;
                }
            }
        }
        s += to_string(3 * r1 + c1 + 1);
        t += d[10 * r2 + c2];
    }

    cout << "Part 1: " << s << endl;
    cout << "Part 2: " << t << endl;

    return 0;
}